### ISSUE TITLE
unconditional_recursion false positive on using function address

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#26165]
This code:

```rust
fn main() {                                                                                                         
    let _test = main as usize;                                                                                      
}                                                                                                                   
```

yields this compiler warning:

```
foo.rs:1:1: 3:2 warning: function cannot return without recurring, #[warn(unconditional_recursion)] on by default   
foo.rs:1 fn main() {                                                                                                
foo.rs:2     let _test = main as usize;                                                                             
foo.rs:3 }                                                                                                          
foo.rs:2:17: 2:21 note: recursive call site                                                                         
foo.rs:2     let _test = main as usize;                                                                             
                         ^~~~                                                                                       
foo.rs:1:1: 3:2 help: a `loop` may express intention better if this is on purpose                                   
```

cc @huonw

**Comment by nagisa** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/21705?

**Comment by alexcrichton** [comment]
Indeed!
